matrix_mqtt_bridge.py

```diff
@@ -70,46 +97,38 @@ def on_publish(client, userdata, mid, properties=None):
 def on_subscribe(client, userdata, mid, granted_qos, properties=None):
     print("[MQTT]", "Subscribed: " + str(mid) + " " + str(granted_qos))
 
+def unescapematch(match):
+    escapesequence = match.group(0)
+    digits = escapesequence[2:]
+    ordinal = int(digits, 16)
+    char = chr(ordinal)
+    return char
 
 # when a MQTT message is recieved
 def on_message(client, userdata, msg):
+    global last_message
     message = msg.payload.decode("utf-8") 
-    print("[MQTT]", msg.topic + " " + str(msg.qos) + " " + str(message))
-
-    global matrix_client, event_loop
-    try:
-        send_fut = asyncio.run_coroutine_threadsafe(
-            matrix_client.room_send(
-                room_id=matrix_room_id,
-                message_type="m.room.message",
-                content={"msgtype": "m.text", "body": message},
-            ),
-            event_loop,
-        )
-        send_fut.result()
-    except nio.exceptions.LocalProtocolError:
-        print("[MTRX]", f"Could not send message to {room_id}, ignoring.")
-
-# using MQTT version 5 here, for 3.1.1: MQTTv311, 3.1: MQTTv31
-# userdata is user defined data of any type, updated by user_data_set()
-# client_id is the given name of the client
-mqtt_client = paho.Client(client_id="Matrix-MQTT-Bridge", userdata=None, protocol=paho.MQTTv31)
```

Comment:
MQTT client setup should only be done once, so maybe we can extract it into it's own function.
```python
async def main() -> None:
    global matrix_client
    matrix_client = AsyncClient(matrix_homeserver, matrix_user)
    matrix_client.add_event_callback(message_callback, RoomMessageText)

    print("[MTRX]", await matrix_client.login(matrix_password))
    # "Logged in as @alice:example.org device id: RANDOMDID"

    await matrix_client.join(matrix_room_id)

    await matrix_client.room_send(
        room_id=matrix_room_id,
        message_type="m.room.message",
        content={"msgtype": "m.text", "body": "[Matrix-MQTT-Bridge]: Ready"},
    )

    await matrix_client.sync_forever(timeout=30000)  # milliseconds

def setup_mqtt():
    # using MQTT version 5 here, for 3.1.1: MQTTv311, 3.1: MQTTv31
    # userdata is user defined data of any type, updated by user_data_set()
    # client_id is the given name of the client
    global mqtt_client
    mqtt_client = paho.Client(callback_api_version=paho.CallbackAPIVersion.VERSION1, client_id="Matrix-MQTT-Bridge", userdata=None, protocol=paho.MQTTv31)
    mqtt_client.on_connect = on_connect

    # enable TLS for secure MQTT connection
    if (mqtt_tls):
        mqtt_client.tls_set(tls_version=mqtt.client.ssl.PROTOCOL_TLS)

    mqtt_client.username_pw_set(mqtt_user, mqtt_password)
    mqtt_client.connect(mqtt_host, int(mqtt_port))

    # setting callbacks, use separate functions like above for better visibility
    mqtt_client.on_subscribe = on_subscribe
    mqtt_client.on_message = on_message
    mqtt_client.on_publish = on_publish

    mqtt_client.loop_start()
```

Then all down at the bottom add the function call:
```python
setup_mqtt()
event_loop = asyncio.new_event_loop()
event_loop.run_until_complete(main())
```

Reply:
But the main is not running in a loop, right? `run_until_complete` already means it's only running once?